content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx

```diff
@@ -0,0 +1,401 @@
+
+### Wrapping the Model in a Class
+
+The most obvious way to wrap the Prisma Client is by wrapping it in a `Model` class:
+
+```tsx highlight=4,23,28-30;normal
+import { PrismaClient } from '@prisma/client'
+
+class Model {
+  constructor(private readonly model: PrismaClient['user']) {}
+
+  /**
+   * Signup the first user and create a new team of one. Return the User with
+   * a full name and without a password
+   */
+  async signup(input: Signup) {
+    assert(input, Signup)
+    const team = await prisma.team.create({
+      data: {
+        ...input.team,
+        users: {
+          create: input.user,
+        },
+      },
+      include: {
+        users: true,
+      },
+    })
+    return exclude(computeFullName(team.users[0]), 'password')
+  }
+}
+
+async function main() {
+  const prisma = new PrismaClient()
+  const model = new Model(prisma.user)
+  const user = await model.signup(/* { ... } */)
+  console.log(user)
+}
+```
+
+Here we're only exposing a `signup` method from the Prisma Client. With this approach, you'll no longer be able to call Prisma Client methods like findMany or upsert. This approach works well when you have a large application and you want to intentionally limit what your models can do.
+
+### Extending the Prisma Client
+
+But what if you don't want to hide existing functionality? In this case, you can use `Object.assign` to extend the Prisma Client without limiting it's functionality:
+
+```tsx highlight=2,27-29;normal
+function Model(model: PrismaClient['user']) {
+  return Object.assign(model, {
+    /**
+     * Signup the first user and create a new team of one. Return the User with
+     * a full name and without a password
+     */
+    async signup(input: Signup) {
+      assert(input, Signup)
+      const team = await prisma.team.create({
+        data: {
+          ...input.team,
+          users: {
+            create: input.user,
+          },
+        },
+        include: {
+          users: true,
+        },
+      })
+      return exclude(computeFullName(team.users[0]), 'password')
+    },
+  })
+}
+
+async function main() {
+  const prisma = new PrismaClient()
+  const model = Model(prisma.user)
+  const user = await model.signup(/* { ... } */)
+  const numUsers = await model.count()
+  console.log(user, numUsers)
+}
+```
+
+Now you can use your custom `signup` method alongside `count`, `updateMany`, `groupBy` and all of the other wonderful methods that the Prisma Client provides. Best of all, it's all type-safe!
+
+Keep in mind that the other methods like `findMany` or `findUnique`, will return users with a `password` and without a `fullName`. If you do go with this approach, you may want to override those functions to modify the results before returning them.
+
+## Conclusion
+
+Today we've shown how you can implement some patterns on top of the Prisma Client to extend functionality and cover your data needs.
```

Comment:
Maybe this could/should be a blog post instead of a docs page? Then a docs page that tells people that this is not possible yet could link to the blog (instead of the Guides section) to find the current workarounds.

Reply:
Ended up writing a page for each. Some copy and paste, but tried to isolate the concepts. I don't have a huge preference on guide vs blog post, just kept it as a guide for now.